### PR TITLE
Refactor user directory tests

### PR DESCRIPTION
**Pull request by DMRobertson**
From #10914 I pulled out changes which refactor the user_directory tests. Suggest reviewing commit-by-commit.

### CODE REVIEW

**Comment by DMRobertson** [comment]
@richvdh hopefully this is easier to digest!

**Comment by DMRobertson** [comment]
CI failing on trial-olddeps.

**Comment by DMRobertson** [comment]
My type hint changes are screwing things up:

```
    from twisted.internet.testing import MemoryReactor
builtins.ModuleNotFoundError: No module named 'twisted.internet.testing'
```

Looks to be using twisted 18.9. [This commit](https://github.com/twisted/twisted/commit/4cade8bb1e5ce45a86962fe2548470a413c8ce7a) moved the location of `MemoryReactor` in twisted 19.7.

File: tests/storage/test_user_directory.py [hunk 1]
```diff
@@ -21,6 +24,39 @@
 BELA = "@somenickname:a"
 
 
+class GetUserDirectoryTables:
+    """Helper functions that we want to reuse in tests/handlers/test+user_directory.py"""
```
Comment (richvdh) [inline tests/storage/test_user_directory.py:33]:
```suggestion
    """Helper functions that we want to reuse in tests/handlers/test_user_directory.py"""
```

File: tests/types.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+try:
+    from twisted.internet.testing import MemoryReactor, MemoryReactorClock
+except ImportError:
+    # Twisted 19.7 moved stuff from twisted.test.proto_helpers to
+    # twisted.internet.testing. Let's do the import once rather than doing it
+    # in lots of different files.
+    from twisted.test.proto_helpers import MemoryReactor, MemoryReactorClock
```
Comment (richvdh) [inline tests/types.py:7]: given it's still available in `twisted.test.proto_helpers` in current twisted, why not just import it from there?

**Review by richvdh — APPROVED** [review]
lgtm other than the possibly over-engineered fix to `MemoryReactor`?

File: tests/types.py (hunk 2 above)
Comment (DMRobertson) [reply tests/types.py:7]: I was leaning towards "make it obvious what the deprecated thing is". No strong opinions though, and we have your CI change to detect any twisted trunk breakages. I'll make that change quickly now.

**Review by richvdh — APPROVED** [review]

**Pull request merged**